File: 0048-rotate-image/0048-rotate-image.cpp
class Solution {
public:
    void rotate(vector<vector<int>>& matrix) {
        int rows = matrix.size();
        int cols = matrix[0].size();

        // Step 1: Transpose the matrix
        for(int i = 0; i < rows; i++) {
            for(int j = i + 1; j < cols; j++) {
                // Swap matrix[i][j] with matrix[j][i]
                int temp = matrix[i][j];
                matrix[i][j] = matrix[j][i];
                matrix[j][i] = temp;
            }
        }

        // Step 2: Reverse the columns (symmetrical column swap)
        for(int i = 0; i < rows; i++) {
            for(int j = 0; j < cols / 2; j++) {
                // Swap the elements in the current row
                int temp = matrix[i][j];
                matrix[i][j] = matrix[i][cols - j - 1];
                matrix[i][cols - j - 1] = temp;
            }
        }
    }
};

File: 0048-rotate-image/0048-rotate-image.py
class Solution:
    def rotate(self, matrix: List[List[int]]) -> None:
        """
        Do not return anything, modify matrix in-place instead.
        """

        #transpose the matrix

        row = len(matrix)
        col = len(matrix[0])

        for i in range(0,row):
            for j in range(i+1, col):
                temp = matrix[i][j]
                matrix[i][j] = matrix[j][i]
                matrix[j][i] = temp
        

        #reverse the columns

        for i in range(0,row):
            fullrow = matrix[i]

            left = 0
            right = len(fullrow) - 1

            while( left <= right):
                temp = fullrow[left]
                fullrow[left] = fullrow[right]
                fullrow[right] = temp
                left += 1
                right -= 1

            matrix[i] = fullrow    